posts/2023-07-18-lazy-layout.lhs

```diff
@@ -0,0 +1,589 @@
+
+> repmin_1pass :: Ord a => Tree a -> Tree a
+> repmin_1pass t = t'
+>  where
+>   (t', globalmin) = repmin t
+>
+>   repmin (Leaf   x)   = (Leaf globalmin, x)
+>   repmin (Branch l r) =
+>     (Branch l' r', min lmin rmin)
+>    where
+>     (l', lmin) = repmin l
+>     (r', rmin) = repmin r
+
+There is an apparent circular dependency here, where `repmin` uses `globalmin`,
+but also computes it.  This is possible because we never need to evaluate
+`globalmin` -- it can be stored as a [thunk].
+For more details, please see the very accessible original paper
+(`https://doi.org/10.1007/BF00264249`).
+
+[thunk]: https://www.youtube.com/watch?v=I4lnCG18TaY
+
+</details>
+
+Starting out with some types
+============================
+
+We start out simple by giving an elegant algebraic definition for a collage:
+
+> data Collage a
+>   = Singleton  a
+>   | Horizontal (Collage a) (Collage a)
+>   | Vertical   (Collage a) (Collage a)
+>   deriving (Foldable, Functor, Show, Traversable)
+
+We will use the [JuicyPixels] library to load and write images.
+The image type in this library can be a bit verbose since it is parameterized
+around the colour space.
+During the layout pass, we don't really care about this, we
+only need the relative sizes of the images and not their content.
+We introduce a typeclass to do just that:
+
+[JuicyPixels]: https://hackage.haskell.org/package/JuicyPixels
+
+> data Size = Size
+>   { sizeWidth  :: Rational
+>   , sizeHeight :: Rational
+>   } deriving (Show)
+>
+> class Sized a where
+>   -- | Retrieve the width and height of an image.
+>   -- Both numbers must be strictly positive.
+>   sizeOf :: a -> Size
+
+We use the `Rational` type for width and height.
+We are only subdividing the 2D space, so we do not need irrational numbers,
+and having infinite precision is convenient.
+
+The instance for the JuicyPixels image type is simple:
+
+> instance Sized (JP.Image p) where
+>   sizeOf img = Size
+>     { sizeWidth  = fromIntegral $ JP.imageWidth  img
+>     , sizeHeight = fromIntegral $ JP.imageHeight img
+>     }
+
+Laying out two images
+=====================
+
+If we look at the finished image, it may seem like a hard problem to find a
+configuration that fits all the images with a correct aspect ratio.
+
+But we can use induction to arrive at a fairly straightforward solution.  Given
+two images, it is always possible to put them beside or above each other by
+scaling them up or down to match them in height or width respectively.  This
+creates a bigger image.  We can then repeat this process until just one image is
+left.
+
+![](/images/2023-07-18-lazy-layout-tree-1.jpg)
+
+However, this is a bit of a naive approach since we end up making way too many
+copies, and the repeated resizing could also result in a loss of resolution.  We
+would like to compute the entire layout first, and then render everything in one
+go.  Still, we can start by formalizing what happens for two images and then
+work our way up.
```

Comment:
```suggestion
However, this is quite a naive approach since we end up making way too many
copies, and the repeated resizing could also result in a loss of resolution.  We
would like to compute the entire layout first, and then render everything in one
go.  Still, we can start by formalizing what happens for two images and then
work our way up.
```